File: LitebaseSDK/src/java/litebase/Key.java
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2020 TotalCross Global Mobile Platform Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

package litebase;

import totalcross.io.*;
import totalcross.util.InvalidDateException;

// juliana@253_5: removed .idr files from all indices and changed its format. 
/**
 * This class represents the key of a record. It may be any of the SQL types defined here.
 */
class Key
{
   /**
    * The size of valRec, an int.
    */
   static final int VALREC_SIZE = 4;

   /**
    * Represents a key that has no values attached to it.
    */
   static final int NO_VALUE = 0xFFFFFFFF; // juliana@230_21

   /**
    * The values stored in the key.
    */
   SQLValue[] keys;

   /**
    * The record index or NO_VALUE.
    */
   int record;

   /**
    * The index that has this key.
    */
   Index index;

   /**
    * Creates a <code>Key</code> object.
    *
    * @param anIndex The index which has the new key.
    */
   Key(Index anIndex)
   {
      // Initializes the key object.
      index = anIndex;
      keys = SQLValue.newSQLValues(index.types.length);
      record = NO_VALUE;
   }

   /**
    * Sets a key of an index.
    *
    * @param key The values used to set the index key.
    */
   void set(SQLValue[] key)
   {
      int i = index.types.length;
      byte[] types = index.types;
      SQLValue[] keysAux = keys;
      
      while (--i >= 0)
      {   
         try
         {
            switch (types[i])
            {
               case SQLElement.DATETIME: // DATETIME.
                  keysAux[i].asShort = key[i].asShort;
               case SQLElement.CHARS: // CHARS and VARCHAR.
               case SQLElement.CHARS_NOCASE: // CHARS NOCASE and VARCHAR NOCASE.
               case SQLElement.DATE: // DATE.
                  keysAux[i].asString = key[i].asString; // juliana@230_3
               case SQLElement.INT: // INT.   
                  keysAux[i].asInt = key[i].asInt;
                  break;
               case SQLElement.SHORT: // SHORT.
                  keysAux[i].asShort = key[i].asShort;
                  break;
               case SQLElement.LONG: // LONG.
                  keysAux[i].asLong = key[i].asLong;
                  break;
               case SQLElement.FLOAT: // FLOAT.
               case SQLElement.DOUBLE: // DOUBLE.
                  keysAux[i].asDouble = key[i].asDouble;                 
               // Blobs can't be used in indices.
            }
         }
         catch (NullPointerException exception) {} // juliana@251_12: removed a possible NPE when using indices with null.
      }
      record = NO_VALUE; // The record key is not stored yet.
   }

   /**
    * Loads a key.
    *
    * @param ds The data stream where the record to be read to find the key value stored.
    * @throws IOException If an internal method throws it.
    */
   void load(DataStreamLB ds) throws IOException, InvalidDateException // juliana@253_8
   {
      Index indexAux = index;
      byte[] types = indexAux.types;
      int n = types.length,
          i = -1;
      int[] colSizes = indexAux.colSizes;
      PlainDB db = indexAux.table.db;
      SQLValue key;
      
      while (++i < n)
      {
         key = keys[i];

         // String keys are not stored in the indices. Only their pointer is stored.
         if (colSizes[i] > 0)
         {
            // If the position is the same, the string is already loaded.
            int pos = ds.readInt();
            if (pos != key.asInt)
            {
               key.asString = null;
               key.asInt = pos;
            }
         }
         else
            // Must pass true to isTemporary so that the method does not think that the number is a rowid.
            // If the value read is null, some bytes must be skipped in the stream.
            // Note: since we're writing only primitive types, we can use any PlainDB available.
            // juliana@220_3 // juliana@230_14
            ds.skipBytes(colSizes[i] - db.readValue(key, 0, types[i], ds, true, false, false)); 
      }
      record = ds.readInt(); // Reads the number that represents the record.
   }

   /**
    * Saves a key.
    *
    * @param ds The data stream where to write the record.
    * @throws IOException If an internal method throws it.
    */
   void save(DataStreamLB ds) throws IOException // juliana@253_8
   {
      Index indexAux = index;
      byte[] types = indexAux.types;
      int n = types.length,
          i = -1;
      int[] colSizes = indexAux.colSizes;
      SQLValue[] keysAux = keys;
      
      while (++i < n)
      {        
         if (colSizes[i] > 0)
            ds.writeInt(keysAux[i].asInt); // Saves only the string position in the .dbo.
         else 
            // If the key is not a string, stores its value in the index file.
            // Note: since primitive types are being written, it is possible to use any PlainDB available.
            // juliana@220_3
            index.table.db.writeValue(types[i], keysAux[i], ds, true, true, 0, 0, false); 
      }
      ds.writeInt(record); // Writes the number that represents the record.
   }
}

File: TotalCrossVM/src/litebase/Key.c
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2020 TotalCross Global Mobile Platform Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

// juliana@253_5: removed .idr files from all indices and changed its format. 
/**
 * Defines functions to deal with the key of a record. It may be any of the SQL types.
 */

#include "Key.h"

/**
 * Sets a key of an index.
 *
 * @param key Yhe the index key to be set.
 * @param SQLValues The values used to set the index key.
 * @param index The index.
 * @param size The number of columns of the index.
 */
void keySet(Key* key, SQLValue** SQLValues, Index* index, int32 size)
{
	TRACE("keySet")
   key->index = index;

	// juliana@202_3: Solved a bug that could cause a GPF when using composed indices.
   while (--size >= 0)
      key->keys[size] = *SQLValues[size];
   key->record = NO_VALUE; // The record key is not stored yet.
}

/**
 * Sets a key using another key.
 *
 * @param to The destination key.
 * @param from The key used to set the other key.
 */
void keySetFromKey(Key* to, Key* from)
{
	TRACE("keySetFromKey")
   Index* index = to->index;
   SQLValue* toKeys = to->keys;
   SQLValue* fromKeys = from->keys;
   SQLValue* fromKey;
   SQLValue* toKey;
   int32* sizes = index->colSizes;
   int32 i = index->numberColumns;
   
   while (--i >= 0)
   {
      fromKey = &fromKeys[i];
      toKey = &toKeys[i];
      if (!sizes[i])
         xmemmove(toKey, fromKey, sizeof(SQLValue)); // full copy
      else
      {
			// juliana@202_8: corrected a bug that would cause string indices to be built incorrectly if they had more than one null value.
			if (fromKey->asChars)
			   xmemmove(toKey->asChars, fromKey->asChars, (toKey->length = fromKey->length) << 1);
			else
            *toKey->asChars = toKey->length = 0;

         toKey->asInt = fromKey->asInt;
      }
   }
   to->index = from->index;
   to->record = from->record;
}

/**
 * Loads a key.
 *
 * @param key The key to be loaded.
 * @param dataStream The data stream where the record to be read to find the key value stored.
 * @return The dataStream offset of the number of bytes read.
 */
uint8* keyLoad(Key* key, uint8* dataStream)
{
	TRACE("keyLoad")
   Index* index = key->index;
   PlainDB* plainDB = &index->table->db;
   SQLValue* keys = key->keys;
   SQLValue* keyAux;
   int8* types = index->types;
   int32* sizes = index->colSizes;
   int32 i = -1,
         n = index->numberColumns,
         pos;

   while (++i < n)
   {
      keyAux = &keys[i];
      if (sizes[i]) // String keys are not stored in the indices. Only their pointer is stored.
      {
         xmove4(&pos, dataStream);
         dataStream += 4;
         if (pos != keyAux->asInt) // If the position is the same, the string is already loaded.
			{
				keyAux->asInt = pos;
				keyAux->length = 0;
				keyAux->asChars[0] = 0;
			}
      }
      else
      {
         // juliana@230_12
         // Must pass true to isTemporary so that the method does not think that the number is a rowid.
         // If the value read is null, some bytes must be skipped in the stream.
         // Note: since we're writing only primitive types, we can use any PlainDB available.
         readValue(null, plainDB, keyAux, 0, types[i], dataStream, true, false, false, -1, null);
         dataStream += typeSizes[types[i]]; 
      }
   }
   xmove4(&key->record, dataStream); // Reads the number that represents the record.
   return dataStream + 4;
}

/**
 * Saves a key.
 *
 * @param key The key to be saved.
 * @param dataStream The data stream where the record to be read to find the key value stored.
 * @return The dataStream offset of the number of bytes written.
 */
uint8* keySave(Key* key, uint8* dataStream)
{
	TRACE("keySave")
   Index* index = key->index;
   SQLValue* keys = key->keys;
   int8* types = index->types;
   int32* sizes = index->colSizes;
	int32 i = -1,
         n = index->numberColumns;

   while (++i < n)
   {
      if (sizes[i]) 
      {
         xmove4(dataStream, &keys[i].asInt); // Saves only the string position in the .dbo.
         dataStream += 4;
      }
      else
      {
         // If the key is not a string, stores its value in the index file.
         // Note: since primitive types are being written, it is possible to use any PlainDB available.
         writeValue(null, null, &keys[i], dataStream, types[i], 0, true, true, false, false);
         dataStream += typeSizes[types[i]];
      }
   }

   xmove4(dataStream, &key->record); // Writes the number that represents the record.
   return dataStream + 4;
}

/**
 * Compares two keys.
 *
 * @param context The thread context where the function is being executed.
 * @param key1 The first key to be compared.
 * @param key2 The second key to be compared.
 * @param isNull1 Indicates if the fist key is null.
 * @param plainDB the plainDB of a table if it is necessary to load a string.
 * @return 0 if the keys are identical; a positive number if <code>key1</code> keys are greater than <code>key2</code> keys; otherwise, a negative 
 * number.
 */
int32 keyCompareTo(Context context, Key* key1, Key* key2, int32 size, PlainDB* plainDB)
{
	TRACE("keyCompareTo")
   int32 r, 
         i = -1;
   int8* types = key1->index->types;
   SQLValue* keys1 = key1->keys;
   SQLValue* keys2 = key2->keys;
   
   while (++i < size) // Compares each key of the key. If a pair is not equal to each other, returns.
      if ((r = valueCompareTo(context, &keys1[i], &keys2[i], types[i], false, false, plainDB)) != 0)
         return r;

   return 0;
}
